janitor/functions/conditional_join.py

```diff
@@ -956,92 +969,104 @@ def _create_frame(
     if not df.columns.intersection(right.columns).empty:
         df, right = _create_multiindex_column(df, right)
 
-    if indicator:
+    def _add_indicator(indicator, how, index_size, nlevels, columns):
+        mapping = {"left": "left_only", "right": "right_only", "inner": "both"}
+        categories = ["left_only", "right_only", "both"]
         if isinstance(indicator, bool):
             indicator = "_merge"
-        if indicator in df.columns.union(right.columns):
+        if indicator in columns:
             raise ValueError(
-                "Cannot use name of an existing column for indicator column"
+                "Cannot use name of an existing column " "for indicator column"
+            )
+        if nlevels > 1:
+            indicator = [indicator] + [""] * (nlevels - 1)
+            indicator = tuple(indicator)
+        if not index_size:
+            arr = pd.Categorical([], categories=categories)
+        else:
+            arr = pd.Categorical(
+                [mapping[how]],
+                categories=categories,
             )
+            if index_size > 1:
+                arr = arr.repeat(index_size)
+        return indicator, arr
+
+    def _inner(df, right, left_index, right_index, indicator):
```

Comment:
Same here for docstrings. Pasting below ChatGPT's version, lightly reviewed by me.

```python
def _inner(
    left: pd.DataFrame,
    right: pd.DataFrame,
    on: Optional[Union[str, List[str], Tuple[str, ...]]] = None,
    left_on: Optional[Union[str, List[str], Tuple[str, ...]]] = None,
    right_on: Optional[Union[str, List[str], Tuple[str, ...]]] = None,
    suffixes: Tuple[str, str] = ("_x", "_y"),
    indicator: Union[bool, str] = False,
    validate: Optional[str] = None,
) -> pd.DataFrame:
    """Perform an inner join between two DataFrames.

    Args:
        left: The left DataFrame to join.
        right: The right DataFrame to join.
        on: Column name or list of column names to join on. If not provided,
            the join will be based on the intersection of columns between
            the two DataFrames.
        left_on: Column name or list of column names from the left DataFrame
            to join on. If provided, it takes precedence over the 'on' parameter.
        right_on: Column name or list of column names from the right DataFrame
            to join on. If provided, it takes precedence over the 'on' parameter.
        suffixes: Suffixes to use for overlapping column names. Default is ('_x', '_y').
        indicator: Add a column to the output DataFrame indicating the source
            of each row. If True, the column name will be "_merge". If a string
            is provided, it will be used as the column name. Default is False.
        validate: If specified, checks if merge is of a specific type.
            Possible values are "one_to_one", "one_to_many", "many_to_one",
            or "many_to_many". Default is None.

    Returns:
        The inner joined DataFrame.

    Raises:
        ValueError: If 'on' is specified and either 'left_on' or 'right_on' is also specified.
        ValueError: If 'on' is not specified and both 'left_on' and 'right_on' are not specified.
        KeyError: If a specified column name is not found in the corresponding DataFrame.

    """
    ...
```

Reply:
chatgpt just copied pandas documentation for me 😁